hw/ip/otbn/util/shared/decode.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import struct
+from typing import Dict, Tuple
+
+from shared.elf import read_elf
+from shared.insn_yaml import Insn, load_insns_yaml
+
+# Load the insns.yml file at module load time.
+try:
+    INSNS_FILE = load_insns_yaml()
+except RuntimeError as err:
+    sys.stderr.write('{}\n'.format(err))
+    sys.exit(1)
+
+
+class OTBNProgram:
+    def __init__(self, symbols: Dict[str, int], insns: Dict[int, int],
+                 data: Dict[int, int]):
+        self.symbols = symbols  # label -> PC
+        self.data = data  # addr -> data (32b word)
+
+        self.insns = {}
+        for pc in insns.keys():
+            opcode = insns[pc]
+            mnem = INSNS_FILE.mnem_for_word(opcode)
+            if mnem is None:
+                raise ValueError('No legal decoding for mnemonic: {}'.format(mnem))
```

Comment:
I'm not sure, but we might plausibly want to pad stuff between functions. I wonder whether it makes sense to just make some sort of illegal instruction object (like the ISS's `IllegalInsn` type) here? (Or `None`, but that makes the types a bit awkward)

Reply:
What do you mean by padding between functions? If the imem has invalid opcodes, isn't that a problem?

Reply:
Well, not necessarily. I mean something like this:
```asm
fun1:
   insn1
   insn2
   insn3
   ret

.zero 20

fun2:
  insn1
  insn2
  insn3
  ret
```
I can't really see a reason we'd need to pad between functions for OTBN (it's not like we've got to worry about cache line alignment), but I also don't think that supporting it is particularly hard.